### DISCUSSION TITLE
Comparing two collections with exact order

### DISCUSSION THREAD

**Discussion by gps7** [nunit/nunit#4448] [How to]
Hi,
I wonder if anyone could suggest how to compare two collections with exactly same order. I know that _Assert.That - Is.EqualTo_ does this correctly with collections, however, this does not provide in special cases a full list of information containing data.

Here is a sample with results:
```
var actual = new[] { "A", "B", "C" };
var expected = new[] { "C", "B", "A" };
Assert.That(actual, Is.EqualTo(expected));
```

Failure gives fallowing information:
  _Expected and actual are both <System.String[3]>
  Values differ at index [0]
  String lengths are both 1. Strings differ at index 0.
  Expected: "C"
  But was:  "A"
  -----------^_

Unfortunately, this does not show full list of actual and expected values which makes very complex to investigate when order is changed since full list is not provided. Trying different approach provides all required information:
```
			var actual = new[] { "A", "B", "C" };
			var expected = new[] { "C", "B", "A" };
			Assert.That(actual, Has.All.EqualTo(expected));
```
Which fails with very useful message:
  _Expected: all items equal to < "C", "B", "A" >
  But was:  < "A", "B", "C" >
  First non-matching item at index [0]:  "A"_

Thanks for comments. If this is not possible with current NUnit version I would suggest a new feature using Assert.That..

**Comment by OsirisTerje** [comment]
If the last one works, what is it you want to be changed?

**Reply by gps7** [reply]
Last one does not work when comparing single item collections:
`Assert.That(new[] { 1 }, Is.All.EqualTo(new[] { 1 }), "Collections are not equal.");`

Resulting in fallowing error:
  _Collections are not equal.
  Expected: all items equal to < 1 >
  But was:  < 1 >
  First non-matching item at index [0]:  1_

**Reply by gps7** [reply]
My assumption that constraint [AllItems ](https://docs.nunit.org/articles/nunit/writing-tests/constraints/AllItemsConstraint.html)meant to be validating each single item instead of my expected intention to compare two collections.

**Comment by OsirisTerje** [comment]
* The first one here looks like a bug.  Please raise this as a bug issue :-) 
* The second , I still don't get it.  "Compare two collections" would be the same as checking that each item match the expected items.

**Reply by gps7** [reply]
Do you mean Is.EqualTo (as in first post) or Is.AllEqualTo with single collection is a bug?

**Reply by gps7** [reply]
With regards second bullet point, I am just expecting print of both collections items in the assert exception by default. The validation works correctly. The workaround would be manually to inject lists into a message:

			var actualString = string.Join(", ", actualNames);
			var expectedString = string.Join(", ", expectedNames);
			Assert.That(actualNames, Is.EqualTo(expectedNames),
				$"Expected collection values mismatched.\r Received: {actualString}\r Expected: {expectedString}");

**Comment by OsirisTerje** [comment]
> Last one does not work when comparing single item collections:

> Assert.That(new[] { 1 }, Is.All.EqualTo(new[] { 1 }), "Collections are not equal.");
> 
> Resulting in fallowing error:
> Collections are not equal.
> Expected: all items equal to < 1 >
> But was: < 1 >
> First non-matching item at index [0]: 1
> 

This is a bug

**Reply by manfred-brands** [reply]
I don't think so.
`All.Equal` I interpret as all items to be the one supplied, not that two arrays match.
The example above should be written as:
```
Assert.That(new[] { 1 }, Is.All.EqualTo(1), "not all elements are 1.");
```
This is not the comparison wanted.
So I think the issue here is to improve the error message for the `Is.EqualTo` case.

**Reply by OsirisTerje** [reply]
@manfred-brands You're right

**Reply by OsirisTerje** [reply]
The first `Assert.That(actual, Has.All.EqualTo(expected));` also then compares each element to the array, and this is more a type mismatch.

**Reply by OsirisTerje** [reply]
The design of the error message is wrt to strings, there it makes sense, it points to the first difference
![image](https://github.com/nunit/nunit/assets/203432/e979bf05-2341-4e5d-9d6c-3be50e4c6892)

It looks like it "think" that a collection is a kind of string, but doesn't report the complete collection.    Adding the AsCollection doesn't change anything.

**Comment by manfred-brands** [comment]
> The first `Assert.That(actual, Has.All.EqualTo(expected));` also then compares each element to the array, and this is more a type mismatch.

Yes, but as everything is an `object`, the compiler doesn't help. Maybe the analyzer could have a rule for this.

**Comment by gps7** [comment]
Based on conversation the Is.EqualTo() is a bug and needs message improvement. Right?
In second part of topic, I would agree to have an analyzer rule, to identify mismatch of types, however, it should consider that array can have arrays and this type of validation will be valid. For example:

```
var actual = new[] { new[] { 1, 2 }, new[] { 2, 1 } };
			var expected = new[] { 1, 2 };
			Assert.That(actual, Is.All.EqualTo(expected));
```

**Comment by gps7** [comment]
Created issue #4450 based on discussion.
